tools/lint.py

```diff
@@ -0,0 +1,127 @@
+
+# Linter script, that calls cpplint.py, specifically for DALI repo.
+# This will be called in `make lint` cmake target
+
+# Q: How to configure git hook for pre-push linter check?
+# A: Create a file `.git/hooks/pre-push`:
+#
+# #!/bin/sh
+# DALI_ROOT_DIR=$(git rev-parse --show-toplevel)
+# python $DALI_ROOT_DIR/tools/lint.py $DALI_ROOT_DIR
+# ret=$?
+# if [ $ret -ne 0 ]; then
+#     exit 1
+# fi
+# exit 0
+
+
+# Specifies, which files are to be excluded
+# These filters are regexes, not typical unix-like path specification
+negative_filters = [
+    ".*core/dynlink_cuda.cc",
+    ".*operators/reader/nvdecoder/nvcuvid.h",
+    ".*operators/reader/nvdecoder/cuviddec.h",
+    ".*operators/reader/nvdecoder/dynlink_nvcuvid.cc",
+    ".*operators/reader/nvdecoder/dynlink_nvcuvid.h",
+    ".*dali/core/dynlink_cuda.h",
+    ".*operators/generic/transpose/cutt/cutt.h",
+    ".*operators/generic/transpose/cutt/cutt.cc",
+    ".*operators/generic/transpose/cutt/cuttplan.h",
+    ".*operators/generic/transpose/cutt/cuttplan.cc",
+    ".*operators/generic/transpose/cutt/cuttkernel.cu",
+    ".*operators/generic/transpose/cutt/cuttkernel.h",
+    ".*operators/generic/transpose/cutt/calls.h",
+    ".*operators/generic/transpose/cutt/cuttGpuModel.h",
+    ".*operators/generic/transpose/cutt/cuttGpuModel.cc",
+    ".*operators/generic/transpose/cutt/cuttGpuModelKernel.h",
+    ".*operators/generic/transpose/cutt/cuttGpuModelKernel.cu",
+    ".*operators/generic/transpose/cutt/CudaMemcpy.h",
+    ".*operators/generic/transpose/cutt/CudaMemcpy.cu",
+    ".*operators/generic/transpose/cutt/CudaUtils.h",
+    ".*operators/generic/transpose/cutt/CudaUtils.cu",
+    ".*operators/generic/transpose/cutt/cuttTypes.h",
+    ".*operators/generic/transpose/cutt/int_vector.h",
+    ".*operators/generic/transpose/cutt/LRUCache.h",
+    ".*python/dummy.cu"
+]
+
+
+def negative_filtering(patterns: list, file_list):
+    """
+    Patterns shall be a list of regex patterns
+    """
+    if len(patterns) == 0:
+        return file_list
+    prog = re.compile(patterns.pop())
+    it = (i for i in file_list if not prog.search(i))
+    return negative_filtering(patterns, it)
+
+
+def gather_files(path: str, patterns: list, antipatterns: list):
+    """
+    Gather files, based on `path`, that match `patterns` unix-like specification
+    and do not match `antipatterns` regexes
+    """
+    curr_path = os.getcwd()
+    os.chdir(path)
+    positive_iterators = [glob.iglob(os.path.join('**', pattern), recursive=True) for pattern in
+                          patterns]
+    linted_files = itertools.chain(*positive_iterators)
+    linted_files = (os.path.join(path, file) for file in linted_files)
+    linted_files = negative_filtering(antipatterns.copy(), linted_files)
+    ret = list(linted_files)
+    os.chdir(curr_path)
+    return ret
+
+
+def gen_cmd(dali_root_dir, file_list, process_includes=False):
+    """
+    Command for calling cpplint.py
+    """
+    return "python " + \
```

Comment:
I wonder if we can run this in multiple threads to make this faster?

Reply:
Good idea, I'll look into it

Reply:
Something like:
```
subproc_num = n
file_list_len = int(math.ceil(len(file_list) / subproc_num)
for i in range(subproc_num):
    if i != subproc_num - 1:
        file_list_slice = file_list[i * file_list_len : (i + 1) * file_list_len]
    else:
         file_list_slice = file_list[i * file_list_len :]
    cmd = "python " + \
          os.path.join(dali_root_dir, "third_party", "cpplint.py") + \
           " --quiet --linelength=100 --root=" + \
           os.path.join(dali_root_dir, "include" if process_includes else "") + \
           " " + ' '.join(file_list_slice )
    process = subprocess.Popen(['ls', '-l'], stdout=subprocess.PIPE)
for i in range(subproc_num):
    stdout, stderr = process.communicate()
    for out in stdout:
        print(stdout)
    for out in stderr :
        print(stdout)
````

Reply:
Done, works sweet now :)